File: src/mcts/node.py
from typing import Optional, Tuple, List
from math import sqrt, log
from src.games.game import Game


class Node:
    def __init__(
        self,
        game: Game,
        parent: Optional["Node"] = None,
        move: Optional[Tuple[int, int]] = None,
    ):
        self.game = game
        self.parent = parent
        self.move = move
        self.children = None
        self.wins = 0
        self.simulations = 0

    def is_fully_explored(self) -> bool:
        return self.children is not None and all(
            child.simulations > 0 for child in self.children
        )

    def is_terminal(self) -> bool:
        return self.game.is_game_over()

    def _uct(self) -> float:
        if self.simulations == 0 or self.parent is None:
            return float("inf")
        exploitation = self.wins / self.simulations
        exploration = sqrt(2 * log(self.parent.simulations) / self.simulations)
        return exploitation + exploration

    def best_child(self) -> "Node":
        if self.children is None:
            return self
        return max(self.children, key=lambda node: node._uct())

    def get_children(self) -> List["Node"]:
        if self.children is None:
            self.children = []
            for move in self.game.get_legal_moves():
                i, j = move
                child = self.game.copy()
                child.make_move(i, j)
                self.children.append(Node(child, self, move))
        return self.children

    def copy(self) -> "Node":
        node = Node(self.game.copy(), self.parent, self.move)
        node.children = self.children
        node.wins = self.wins
        node.simulations = self.simulations
        return node

File: tests/java/agents/Node.java
package tests.java.agents;

import java.util.ArrayList;
import tests.java.games.BoardState;
import tests.java.games.Move;

public class Node {

    private BoardState state;
    private Node parent;
    private Move move;
    private ArrayList<Node> children;
    private double value;
    private int visits;

    public Node(BoardState state, Node parent, Move move) {
        this.state = state;
        this.parent = parent;
        this.move = move;
        this.children = null;
        this.value = 0;
        this.visits = 0;
    }

    public Node(Node copyNode) {
        this.state = copyNode.state.copy();
        this.parent = copyNode.parent;
        this.move = copyNode.move;
        this.children = null;
        this.value = copyNode.value;
        this.visits = copyNode.visits;
    }

    public void incrementVisits(int count) {
        this.visits += count;
    }

    // Increment wins
    public void incrementValue(double value) {
        this.value += value;
    }

    public boolean isFullyExpanded() {
        return (
            children != null &&
            children.stream().allMatch(child -> child.visits > 0)
        );
    }

    public boolean isTerminal() {
        return state.isGameOver();
    }

    private double uct(double c) {
        if (visits == 0 || parent == null) {
            return Double.POSITIVE_INFINITY;
        }
        double exploitation = value / visits;
        double exploration = c * Math.sqrt(Math.log(parent.visits) / visits);
        return exploitation + exploration;
    }

    public Node bestChild(double c) {
        if (children == null || children.isEmpty()) {
            return this; // No children; return self
        }
        return children
            .stream()
            .max((child1, child2) ->
                Double.compare(child1.uct(c), child2.uct(c))
            )
            .orElse(this);
    }

    public ArrayList<Node> getChildren() {
        if (children == null) {
            children = new ArrayList<>();
            for (Move legalMove : state.getLegalMoves()) {
                BoardState newState = state.copy();
                newState.makeMove(legalMove);
                children.add(new Node(newState, this, legalMove));
            }
        }
        return children;
    }

    public BoardState getState() {
        return state;
    }

    public Node getParent() {
        return parent;
    }

    public Move getMove() {
        return move;
    }

    public double getValue() {
        return value;
    }

    public void setValue(double value) {
        this.value = value;
    }

    public int getVisits() {
        return visits;
    }

    public void setVisits(int visits) {
        this.visits = visits;
    }
}
